src/backend/distributed/replication/multi_logical_replication.c

```diff
@@ -349,6 +369,243 @@ LogicallyReplicateShards(List *shardList, char *sourceNodeName, int sourceNodePo
 }
 
 
+/*
+ * InitPublicationInfoHash initializes an empty hash map which can be used to
+ * store PublicationInfo entries.
+ */
+HTAB *
+InitPublicationInfoHash(void)
+{
+	HASHCTL info;
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(NodeAndOwner);
+	info.entrysize = sizeof(PublicationInfo);
+	info.hash = HashNodeAndOwner;
+	info.match = CompareNodeAndOwner;
+	info.hcxt = CurrentMemoryContext;
+
+	int hashFlags = (HASH_ELEM | HASH_CONTEXT | HASH_FUNCTION | HASH_COMPARE);
+
+	HTAB *publicationInfoHash = hash_create("PublicationInfoHash", 128, &info, hashFlags);
+	return publicationInfoHash;
+}
+
+
+/*
+ * InitNodeSubscriptionsHash initializes an empty hash map which can be used to
+ * store the SubscriptionInfo entries grouped by node.
+ */
+static HTAB *
+InitNodeSubscriptionsHash(void)
+{
+	HASHCTL info;
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(uint32);
+	info.entrysize = sizeof(NodeSubscriptions);
+	info.hash = HashNodeId;
+	info.match = CompareNodeId;
+	info.hcxt = CurrentMemoryContext;
+
+	int hashFlags = (HASH_ELEM | HASH_CONTEXT | HASH_FUNCTION | HASH_COMPARE);
+
+	HTAB *publicationInfoHash = hash_create("NodeSubscriptionsHash", 128, &info,
+											hashFlags);
+	return publicationInfoHash;
+}
+
+
+/*
+ * SubscriptionInfoNodeIdHash returns hash value by combining hash of node id
+ * and tableowner Id.
+ */
+static uint32
+HashNodeId(const void *key, Size keysize)
+{
+	uint32 *nodeId = (uint32 *) key;
+	return hash_uint32(*nodeId);
+}
+
+
+/*
+ * NodeAndOwnerHash returns hash value by combining hash of node id
+ * and tableowner Id.
+ */
+uint32
+HashNodeAndOwner(const void *key, Size keysize)
```

Comment:
do we need explicit hash functions? seems like tag_hash would do fine, meaning you can use HASH_BLOBS flag and not specify a hash function.

Reply:
(default compare function is memcp)

Reply:
I also kind of wonder whether we could have some more generic CreateHashTable function.

Reply:
Cool, didn't know about this way of creating HTABs. Done now.

Reply:
> I also kind of wonder whether we could have some more generic CreateHashTable function.

You ask I deliver